File: 503. Next Greater Element II/Solution.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        newNums = nums + nums[:-1]
        result = []
        mapping = collections.defaultdict(list)
        stack= [nums[0]]
        
        for i in range(1, len(newNums)):
            while stack and stack[-1] < newNums[i]:
                mapping[stack[-1]].append(newNums[i])
                stack.pop()
            stack.append(newNums[i])
            
        for el in nums:
            if el in mapping:
                result.append(mapping[el].pop(0))
            else:
                result.append(-1)
        
        return result

File: 503. Next Greater Element II/Solution.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] index = new int[n];
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 2 * n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && nums[i % n] >= stack.peek()) {
                stack.pop();
            }
            index[i % n] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(nums[i % n]);
        }
        
        return index;
    }
}